src/live.rs

```diff
@@ -456,9 +457,76 @@ struct KargEmbedAreas {
     args: String,
 }
 
+#[derive(Deserialize)]
+struct KargEmbedInfo {
+    default: String,
+    files: Vec<KargEmbedLocation>,
+    size: usize,
+}
+
+#[derive(Deserialize)]
+struct KargEmbedLocation {
+    path: String,
+    offset: u64,
+}
+
 impl KargEmbedAreas {
     // Return Ok(None) if no kargs embed areas exist.
-    pub fn for_file(file: &mut File) -> Result<Option<Self>> {
+    pub fn for_iso(iso: &mut IsoFs) -> Result<Option<Self>> {
+        let iso_file = match iso.get_path(COREOS_KARG_EMBED_INFO_PATH) {
+            Ok(record) => record.try_into_file()?,
+            // old ISO without info JSON
+            Err(e) if e.is::<iso9660::NotFound>() => {
+                return Self::for_file_via_system_area(iso.as_file()?)
+            }
+            Err(e) => return Err(e),
+        };
+        let info: KargEmbedInfo = serde_json::from_reader(
+            iso.read_file(&iso_file)
+                .context("reading kargs embed area info")?,
+        )
+        .context("decoding kargs embed area info")?;
+
+        // sanity-check size against a reasonable limit
+        if info.size > COREOS_KARG_EMBED_AREA_MAX_SIZE {
+            bail!(
+                "karg embed area size larger than {} (found {})",
+                COREOS_KARG_EMBED_AREA_MAX_SIZE,
+                info.size
+            );
+        }
+        if info.default.len() > info.size {
```

Comment:
Hmm, so then do we even need a `size` field in the JSON file given that the information is already provided via the string length itself?

Reply:
Unlike in the old `.cmdline` file, the `default` string is not padded to the size of the embed area, so the sizes are in fact different.